{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ebd57fc-494f-4811-85d6-1960af008715",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in c:\\users\\steff\\anaconda3\\lib\\site-packages (1.85)\n",
      "Requirement already satisfied: numpy in c:\\users\\steff\\anaconda3\\lib\\site-packages (from biopython) (1.26.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0875f94d-6987-4a10-b868-b3b83b278414",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Installing Bioython\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d933a5d3-a229-40e8-bdd1-b1d49c109467",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTACATGGTA\n",
      "TCATGTACCAT\n",
      "TACCATGTACT\n",
      "AGUACAUGGUA\n",
      "STW\n"
     ]
    }
   ],
   "source": [
    "my_seq = Seq(\"AGTACATGGTA\")\n",
    "print(my_seq)\n",
    "print(my_seq.complement())\n",
    "print(my_seq.reverse_complement())\n",
    "print(my_seq.transcribe())\n",
    "print(my_seq.translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a60e744-52fe-44be-b26e-ce4bffc76d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse and read FASTA using BioPython\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6ccbf0b9-b1f8-4712-88a6-9573bdd78917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NW_021636346.1:c40256-27400\n",
      "Name: NW_021636346.1:c40256-27400\n",
      "Description: NW_021636346.1:c40256-27400 LOC114717194 [organism=Prosopis alba] [GeneID=114717194] [chromosome=Un]\n",
      "Number of features: 0\n",
      "Seq('TCACGGAGCTTAAACAAGGCCCTCTATAGCCTCTGTCTGCTTACTGTGCCACAA...AAA')\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.read(\"C:/Users/steff/Documents/Data Sciences For Biologist/BioPython-FASTA/gene.fna\", \"fasta\") \n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b50f3b0-a45c-4909-bef1-1b48fa03a1e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NW_021636346.1:c40256-27400'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c30d9f04-a8fd-46df-9e7a-f6cab089762c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NW_021636346.1:c40256-27400 LOC114717194 [organism=Prosopis alba] [GeneID=114717194] [chromosome=Un]'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bede2f3-c71f-4c39-a05e-6537ce2516f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "74a3f383-de22-4c48-905f-d24f3e25e36f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence (first 100 bases): TCACGGAGCTTAAACAAGGCCCTCTATAGCCTCTGTCTGCTTACTGTGCCACAAAAACGCCCACCTTCATAACGGCTTCTTGCATAAGCGACTGTGGAAG\n"
     ]
    }
   ],
   "source": [
    "len(record)\n",
    "print(f\"Sequence (first 100 bases): {record.seq[:100]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dfb82a17-bbe6-4eae-837f-d0504fe4432c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Writing a FASTA file\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "seq_obj = SeqRecord(Seq(\"ATGCATGACTGA\"), id = \"Test\",\n",
    "                    description = \"Example sequence\")\n",
    "SeqIO.write(seq_obj, \"output.fasta\", \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7c4f676e-813a-4852-9895-31cab9ea0f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAIM*SAGGPH\n"
     ]
    }
   ],
   "source": [
    "# Practical Use Case - Sequence Translation\n",
    "dna_seq =Seq(\"ATGGCCATTATGTAATCCGCGGGTGGCCCGCATAG\")\n",
    "protein_seq = dna_seq.translate()\n",
    "print(protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "67df67c6-0010-44e3-95b3-3d6c673e3807",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NW_021636346.1:c40256-27400: 37.58%GC\n"
     ]
    }
   ],
   "source": [
    "# To calculate the GC count\n",
    "from Bio import SeqIO\n",
    "\n",
    "def calculate_gc(seq):\n",
    "    g = seq.count(\"G\")\n",
    "    c = seq.count(\"C\")\n",
    "    gc_content = 100*float(g+c)/len(seq)\n",
    "    return gc_content\n",
    "\n",
    "for record in SeqIO.parse(\"gene.fna\", \"fasta\"):\n",
    "    print(f\"{record.id}: {calculate_gc(record.seq):.2f}%GC\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b06ff43d-3d88-40b6-b710-c1f2e8a50aab",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
